router deployment

jlarriba opened the pull request.

jlarriba commented:
I have not been able to test this yet, so do not merge. But I think it is interesting to have it visible for closer collaboration during ShiftWeek

pierreprinetti commented on .gitignore:
```diff
@@ -24,5 +24,6 @@ Dockerfile.cross
 *.swo
 *~
+
+# secrets
```
Strange. This change is in `main` already

jlarriba replied:
I created my branch on monday, and merged main changes yesterday. Maybe it is because of this.

pierreprinetti commented on api/v1alpha1/openstackrouter_types.go:
```diff
@@ -0,0 +1,177 @@
+
+// GatewayInfo represents the information of an external gateway for any
+// particular network router.
+type GatewayInfo struct {
```
Please namespace this type under the name of the resource:

```suggestion
type OpenStackRouterGatewayInfo struct {
```

pierreprinetti commented on api/v1alpha1/openstackrouter_types.go:
```diff
@@ -0,0 +1,177 @@
+
+// ExternalFixedIP is the IP address and subnet ID of the external gateway of a
+// router.
+type ExternalFixedIP struct {
```
Please namespace this type under the name of the resource:

```suggestion
type OpenStackRouterExternalFixedIP struct {
```

pierreprinetti commented on api/v1alpha1/openstackrouter_types.go:
```diff
@@ -0,0 +1,177 @@
+}
+
+// Route is a possible route in a router.
+type Route struct {
```
Please namespace this type under the name of the resource:

```suggestion
type OpenStackRouterRoute struct {
```

pierreprinetti commented on api/v1alpha1/openstackrouter_types.go:
```diff
@@ -0,0 +1,177 @@
+// InterfaceInfo represents information about a particular router interface. As
+// mentioned above, in order for a router to forward to a subnet, it needs an
+// interface.
+type InterfaceInfo struct {
```
Please namespace this type under the name of the resource:

```suggestion
type OpenStackRouterInterfaceInfo struct {
```

pierreprinetti commented on api/v1alpha1/openstackrouter_types.go:
```diff
@@ -0,0 +1,177 @@
+
+	Description string `json:"description,omitempty"`
+
+	ExternalGateway string `json:"externalGateway,omitempty"`
```
This is called `external_gateway_info` in the API [docs](https://docs.openstack.org/api-ref/network/v2/#create-router). It should be `externalGatewayInfo` here for consistency

jlarriba replied:
That's strange. What the API is referring to is the ExternalGatewayInfo structure that contains a set of things. The externalGateway here is just a kubernetes name reference to a network, not an actual external_gateway_info object.

It tries to mimick the parameter that is passed on the CLI: _openstack router create vrouter --external-gateway public_

pierreprinetti replied:
what I'd do, is that I'd run the same openstackclient command with `--debug` to see how `public` is passed

jlarriba replied:
I have just tried it and it does the same that the controller does: it retrieves the network by name and wraps the NetworkID inside an external_gateway_info. 

The wrapper: ttps://github.com/gophercloud/gopherkube/pull/47/files#diff-1b8b81f0722366512bb81963c60c8f7f2fed15d6f3a42d8f3ad42489d62ff43eR168

pierreprinetti replied:
Perfect. Then I'd expose here the same struct that is exposed in the API, leaving the wrapping to the user

pierreprinetti commented on api/v1alpha1/openstackrouter_types.go:
```diff
@@ -0,0 +1,177 @@
+	ExternalGateway string `json:"externalGateway,omitempty"`
+
+	// AdminStateUp is the administrative state of the router.
+	AdminStateUp *bool `json:"adminStateUp,omitempty"`
```
The API [docs](https://docs.openstack.org/api-ref/network/v2/#create-router) declares the default of this field to be `true`. Based on this, we can remove the pointer and make the value default to `true`:

```suggestion
	// +kubebuilder:default:=true
	AdminStateUp bool `json:"adminStateUp,omitempty"`
```

pierreprinetti replied:
The thing is, defaults sometimes change. Maybe your pointer solution was more future-proof. What do you think?

pierreprinetti replied:
After some thought, your pointer solution was more consistent with the API. Like the API, it lets you pick one of three choices: true, false or "you pick". Can you please revert to pointer and excuse me for the back-and-forth?

jlarriba replied:
Sure! no worries

pierreprinetti reviewed (commented):
Thank you for this work! I am posting a couple initial comments.

jlarriba commented:
/retest

jlarriba commented:
It seems that https://github.com/gophercloud/gopherkube/commit/5f74be53df963e4038280384a236397bafb01236 broke the testing environment

pierreprinetti commented:
> It seems that [5f74be5](https://github.com/gophercloud/gopherkube/commit/5f74be53df963e4038280384a236397bafb01236) broke the testing environment

yes it did unfortunately. At a minimum, `clouds.NewClient` should be replaced with `clouds.NewServiceClient` (the Gophercloud generation was split in two).

vkmc reviewed (approved):
Looks good and it's working as expected. Just left a couple of comments, but it's minor stuff. For me it's good to go :) Thanks Juan!

jlarriba commented on api/v1alpha1/openstackrouter_types.go:
```diff
@@ -0,0 +1,185 @@
+
+	// TenantID is the project owner of the router. Only admin users can
+	// specify a project identifier other than its own.
+	TenantID string `json:"tenantID,omitempty"`
```
I dont think so, it is implemented by the Openstack API, so I think it should be there.

pierreprinetti commented on api/v1alpha1/openstackrouter_types.go:
```diff
@@ -0,0 +1,185 @@
+
+	Description string `json:"description,omitempty"`
+
+	ExternalGatewayInfo OpenStackRouterExternalGatewayInfo `json:"externalGatewayInfo,omitempty"`
```
This is optional. The type should be a pointer, so that it's easier to know whether the user actually passed one

```suggestion
	ExternalGatewayInfo *OpenStackRouterExternalGatewayInfo `json:"externalGatewayInfo,omitempty"`
```

pierreprinetti replied:
:point_up:

jlarriba replied:
Yes, I am working on a PR to solve some issues at the same time.

pierreprinetti commented on api/v1alpha1/openstackrouter_types.go:
```diff
@@ -0,0 +1,185 @@
+	ProjectID string `json:"projectID,omitempty"`
+
+	// All the subnets that will be added as router interfaces
+	Subnets []string `json:"subnets,omitempty"`
```
I've been thinking a lot about this one. I agree that we must provide a way to add interfaces right from this spec... the problem now is _how_.

As far as I know, there are two ways to add an interface to a router:
* call `add_router_interface` with a `subnet_id` (which is what you do here); OR
* create a port and call `add_router_interface` with its `port_id`.

I think we should only provide a way to adding ports, and not subnets, to routers. The reason for this is: when you pass a subnet, Neutron automatically creates a port, which is not represented anywhere in our CRs. Instead if you create a port (for which we will get a CRD soon) and only then add it to a router, then everything is nicely accounted for in our YAMLs.

By adding `Ports` you enable users to add a specific IP to their router, while with `Subnets` you only allow ports bound to the subnet's gateway IP. In that respect, passing a `port_id to `add_router_interface` is a strict superset of passing a `subnet_id`.

---

My proposal: remove `Subnets` and add `Port` instead... which accepts a port ID until we add `OpenStackPort` to the CRDs.

---

What do you think?

jlarriba replied:
The current implementation was done taking in account the fact that we were developing the Router before the Port, so only one way of adding interfaces could exist (via Subnet).

Having that in mind I think is that it is much more intuitive to add Subnets directly than having to create a Port. The port that Neutron creates is included in the interfacesinfo. What we could do is to create Port CRs when the information from Neutron comes to have them represented in our CRs.

Of course, providing a Port should also be supported when it is possible, but I think the use case of providing a Subnet would be principal one.

Maybe it would be a good approach to merge this with Subnet support and then contribute a new PR adding the support to Ports when that is included (at least, that was my intention).

pierreprinetti reviewed (changes requested):
Juan, congratulations, you didn't pick the simplest resource :joy: 

This review took me a while. I have comments on the API part; I have postponed the review of the controller until we converge to an API. Please let me know what you think.

jlarriba commented on api/v1alpha1/openstackrouter_types.go:
```diff
@@ -0,0 +1,185 @@
+	GatewayInfo OpenStackRouterGatewayInfo `json:"externalGatewayInfo,omitempty"`
+
+	// InterfacesInfo provides information on the interfaces connected to this router
+	InterfacesInfo []OpenStackRouterInterfaceInfo `json:"interfacesInfo,omitempty"`
```
It is not a list of ports, it is a list of a combination of port and subnet

pierreprinetti commented:
Hey @jlarriba ! Do you mind squashing the commits? At this stage of the project, it's probably best to merge and iterate. I am going to merge a big refactoring and I don't want you to have to resolve 1000 conflicts

jlarriba commented:
Hi, I  had the changes addressing the comments lingering in local, I thought I had pushed them :( Solved all the smaller things and squashed the commits.

jlarriba commented on config/crd/kustomization.yaml:
```diff
@@ -11,6 +12,18 @@ resources:
 - bases/openstack.k-orc.cloud_openstacksecuritygroups.yaml
 - bases/openstack.k-orc.cloud_openstackservers.yaml
 - bases/openstack.k-orc.cloud_openstacksubnets.yaml
+=======
```
Sorry, one cannot be at 4 things at the same time.

pierreprinetti commented on config/crd/kustomization.yaml:
```diff
@@ -2,15 +2,16 @@
+- bases/openstack.gopherkube.dev_openstacksecuritygroups.yaml
+- bases/openstack.gopherkube.dev_openstackservers.yaml
+- bases/openstack.gopherkube.dev_openstacksubnets.yaml
+- bases/openstack.gopherkube.dev_openstackrouters.yaml
```
nope xD

jlarriba replied:
OMG!

pierreprinetti reviewed (approved):
Thank you!
